dpnp/dpnp_utils/dpnp_utils_linearalgebra.py

```diff
@@ -1894,108 +2024,99 @@ def dpnp_matmul(
+        call_flag = "gemm_batch"
+        res_shape = result_shape
+    elif x1_is_1D and x2_is_2D:
+        # TODO: implement gemv to use it here with transpose
+        call_flag = "gemm"
+        x1 = dpnp.reshape(x1, (1, x1.size))
+        x2 = dpnp.reshape(x2, x2_shape[-2:])
         x1_shape = x1.shape
+        res_shape = (x1_shape[-2], x2_shape[-1])
+    elif x1_is_2D and x2_is_1D:
+        # TODO: implement gemv to use it here without transpose
+        call_flag = "gemm"
+        x1 = dpnp.reshape(x1, x1_shape[-2:])
+        x2 = dpnp.reshape(x2, (x2.size, 1))
         x2_shape = x2.shape
-        if out is not None:
-            for i in range(x1_ndim - 2):
-                if tmp_shape[i] != out_shape[i]:
-                    if not appended_axes:
-                        _shape_error(tuple(tmp_shape), out_shape[:-2], None, 3)
-                    elif len(appended_axes) == 1:
-                        _shape_error(tuple(tmp_shape), out_shape[:-1], None, 3)
-
-        res_shape = tuple(tmp_shape) + (x1_shape[-2], x2_shape[-1])
-
-    # handling a special case to provide a similar result to NumPy
-    if out is not None and x1.shape == (1, 0) and x2.shape == (0, 1):
-        res_shape = (0,)
-        appended_axes = []
-
-    result = _create_result_array(
-        x1, x2, out, res_shape, gemm_dtype, res_usm_type, exec_q
-    )
+        res_shape = (x1_shape[-2], x2_shape[-1])
+    elif x1_is_2D and x2_is_2D:
+        call_flag = "gemm"
+        x1 = dpnp.reshape(x1, x1_shape[-2:])
+        x2 = dpnp.reshape(x2, x2_shape[-2:])
+        res_shape = (x1_shape[-2], x2_shape[-1])
+    elif x1_base_is_1D:
+        # TODO: implement gemv_batch to use it here with transpose
+        call_flag = "gemm_batch"
+        if x1_ndim == 1:
+            x1 = dpnp.reshape(x1, (1, 1, x1.size))
+            res_shape = result_shape[:-1] + (1, result_shape[-1])
+        else:
+            res_shape = result_shape
+    elif x2_base_is_1D:
+        # TODO: implement gemv_batch to use it here without transpose
+        call_flag = "gemm_batch"
+        if x2_ndim == 1:
+            x2 = dpnp.reshape(x2, (1, x2.size, 1))
+            res_shape = result_shape + (1,)
+        else:
+            res_shape = result_shape
+    else:
+        call_flag = "gemm_batch"
+        res_shape = result_shape
+
+    if call_flag == "kron":
+        res = dpnp.kron(x1, x2)
+        res_shape = res.shape
+    elif call_flag == "dot":
+        if out is not None and out.shape != ():
+            res = dpnp_dot(x1, x2)
+        else:
+            res = dpnp_dot(x1, x2, out=out)
+        res_shape = res.shape
+    else:
+        res = _create_result_array(
+            x1, x2, out, res_shape, compute_dtype, res_usm_type, exec_q
+        )
 
     # calculate result
-    if result.size == 0:
+    if res.size == 0:
         pass
     elif x1.size == 0 or x2.size == 0:
-        result.fill(0)
+        res.fill(0)
     else:
```

Comment:
I guess we should only get here when `call_flag not in ("kron", "dot")`, i.e. we should tab the code between lines#2115-2169 to have it under the else block at line#2110.